### PR TITLE
Tag structs as atomic

### LINKED ISSUE #97970: Server Side Apply - Selectors are not atomic
While closing the https://github.com/kubernetes/kubernetes/issues/92913 fixed a problem for structured selectors across `k/k` api, there is still a great deal of unstructured selectors across different resources as a `map[string]string`. So a similar problem is present - SSA does not work for modifying these when a multiple source-of-truths attempting to manage this field in parallel. 

Say a user and a controller is doing changes to the resource.selector field. If a user changes (appends) a key: value to the map, he completely alters the behavior for the resource, and the controller can't do anything with it, as SSO treats the map as `granular` resource by default, so you have to fall back to a usual `get/compare/update` for changes like that, or you can't guarantee the declarative behavior for your controller, which is supposed to set and maintain the fields it cares about as it expects it to see.

**What happened**:
Here is a resource:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: test
  namespace: default
spec:
  ports:
...
  selector:
    test: test
```
1) User uses `kubectl apply --server-side -f svc.yaml`, the resource is applied.
2) A controller for this resource reconciles its changes, and applies a patch with
```yaml
apiVersion: v1
kind: Service
...
  selector:
    my: app
```
3) The resource contains both values in `spec.selector` map and fails to do the job with existing infrastructure, as the selector does not match any of the resource labels.
```yaml
...
  selector:
    my: app
    test: test
```

**What you expected to happen**:

Step 2 applies a selector as an atomic value, so the resource in step 3 looks more like
```yaml
...
  selector:
    my: app
```

**How to reproduce it (as minimally and precisely as possible)**:

See **what happend** section

A followup to https://github.com/kubernetes/kubernetes/issues/92913

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.20
- Cloud provider or hardware configuration: does not matter

### PR DESCRIPTION
**Pull request by Jefftree**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
See [review of all occurrences of the term "selector" in types.go files](https://docs.google.com/document/d/1jA39N3uvyDHheE-g8nLKbbnKkB3zhe4plAyl8qSdk_k/edit#heading=h.6e6lmpudp0ay)

Addresses comments from https://github.com/kubernetes/kubernetes/pull/100131

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/97970

Related to https://github.com/kubernetes/kubernetes/pull/97989

#### Special notes for your reviewer:

I'm not sure what kind of tests we need for these. There are already tests for the generic atomic annotation, I feel like we shouldn't need to test every instance of it. Some are transparent changes because the types are only used in lists that are already atomic.

#### Does this PR introduce a user-facing change?

```release-note
SSA treats certain structs as atomic
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/555-server-side-apply
- Feature Issue: kubernetes/enhancements#555
- List of "selector" occurrences in types.go files: https://docs.google.com/document/d/1jA39N3uvyDHheE-g8nLKbbnKkB3zhe4plAyl8qSdk_k/edit

### CODE REVIEW

**Comment by apelisse** [comment]
/lgtm

Thanks, I agree I don't think we should add tests for each of these.

**Comment by apelisse** [comment]
Can you maybe squash some of these commits?

**Comment by Jefftree** [comment]
Squashed

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -2668,7 +2674,10 @@ const (
 // The requirements of them are ANDed.
 // It provides a subset of functionality as NodeSelectorTerm.
 // This is an alpha feature and may change in the future.
+// +structType=atomic
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:2677]: should NodeSelectorTerm be marked atomic as well?

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2668,7 +2674,10 @@ const (
...
 type TopologySelectorTerm struct {
+	// Usage: Fields of type []TopologySelectorTerm must be listType=atomic.
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:2679]: There are `AllowedTopologies []v1.TopologySelectorTerm` uses that are not marked as listType=atomic... where will that be done?

Comment (Jefftree) [reply staging/src/k8s.io/api/core/v1/types.go:2679]: Commented on Joe's PR earlier, listType is atomic by default so omitting the annotation doesn't change any behavior. Would you prefer it to be explicitly marked?

Comment (apelisse) [reply staging/src/k8s.io/api/core/v1/types.go:2679]: I'm personally happy with explicitly marking it as atomic

**Comment by Jefftree** [comment]
Addressed comments. Marked `AllowedTopologies []v1.TopologySelectorTerm` usages with listType=atomic.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 3]
```diff
@@ -5362,6 +5374,7 @@ type ServiceProxyOptions struct {
 // Instead of using this type, create a locally provided and used type that is well-focused on your reference.
 // For example, ServiceReferences for admission registration: https://github.com/kubernetes/api/blob/release-1.17/admissionregistration/v1/types.go#L533 .
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +structType=atomic
 type ObjectReference struct {
```
[the remark this answers is not in the document]
Comment (apelisse) [reply staging/src/k8s.io/api/core/v1/types.go:5378]: Good catch Janet, thanks for letting us know. @alexzielenski is looking at it.

[the remark this answers is not in the document]
Comment (apelisse) [reply staging/src/k8s.io/api/core/v1/types.go:5378]: Related too, unsurprisingly: https://github.com/fluxcd/flux2/issues/2827

**Pull request merged**